Python/pystate.c

```diff
@@ -1057,6 +1057,47 @@ _PyErr_SetInterpreterAlreadyRunning(void)
     PyErr_SetString(PyExc_InterpreterError, "interpreter already running");
 }
 
+int
+_PyInterpreterState_IsShuttingDown(PyInterpreterState *interp)
+{
+    return get_main_thread(interp) == _PyInterpreterState_MAIN_SHUTTING_DOWN;
+}
+
+int
+_PyInterpreterState_SetShuttingDown(PyInterpreterState *interp)
+{
+    void *expected = NULL;
+    // XXX Use an exchange without a comparison?
+    if (_Py_atomic_compare_exchange_ptr(&interp->threads.main,
+                                        &expected,
+                                        _PyInterpreterState_MAIN_SHUTTING_DOWN) == 0)
+    {
+        /* We're either already shutting down or another thread started running */
+        _PyErr_SetInterpreterAlreadyRunning();
+        return -1;
+    }
+    assert(_PyInterpreterState_IsShuttingDown(interp));
+    assert(!_PyInterpreterState_IsRunningMain(interp));
+    /* At this point, we're certain that no other threads can set the main thread.
+     * However, there might be some remaining thread states--in that case, let's just raise.
+     * We could wait for them all to finish up, but that's a job for later. */
```

Comment:
```suggestion
    /* At this point, we're certain that no other threads can set the main thread.
     * However, there might be some remaining thread states--in that case, let's just raise.
     * /
```
I don't think it's necessary to explain the future work.

Reply:
It is, there's plenty of XXX and TODO comments throughout the subinterpreter codebase.